Homework for sprint №8

sobolevkir opened the pull request:
In this change, the main and the settings screens have been implemented, as well as navigation.

WARNING! Please pay attention to the layout, I would like to receive comments on it, especially on the implementation of margins and basic views. Thank you!

sobolevkir committed:
Main screen and settings screen have been created.

sobolevkir committed:
Clicks on the home screen buttons have been implemented

sobolevkir committed:
Navigation between screens have been implemented

AleksandrIlinskii commented on app/src/main/java/com/sobolevkir/playlistmaker/MainActivity.kt:
```diff
@@ -1,11 +1,39 @@
+        }
+
+        //Нажатие на третью кнопку реализую через анонимный класс
+        val settingsButtonClickListener: View.OnClickListener = object : View.OnClickListener {
```
Супер, что ты знаешь как это делается 😄 
Но обычно так никто не делает, так что в следующем спринте удали пожалуйста

sobolevkir replied:
Там было написано, что двумя способами надо сделать, а после этого спринта заменить на лямбда)

AleksandrIlinskii commented on app/src/main/java/com/sobolevkir/playlistmaker/SettingsActivity.kt:
```diff
@@ -0,0 +1,20 @@
+        val backButton = findViewById<ImageButton>(R.id.button_back)
+
+        backButton.setOnClickListener{
+            startActivity(Intent(this, MainActivity::class.java))
```
Здесь лучше сделать finish(). Активити уничтожается и ты возвращаешься обратно по стеку на главную. 
В данном случае что происходит - ты накладываешь как пирог
главная -> настройки -> главная -> настройки
Чтобы воспроизвести проблему, сделай вот как - с главной нажми на настройки. Потом на твою стрелку назад. Потом снова на настройки, снова назад и на настройки. А теперь нажимай не на стрелку назад, а на системную кнопку назад. Попадешь на главную. Но что должно произойти когда ты нажмешь назад? Правильно, должно выйти. А что будет? Ты попадешь на настройки, потому что они последующие в стеке

sobolevkir replied:
А finish куда нужно вписать?

sobolevkir replied:
Кстати, в теории этого не было вроде, можешь поподробнее пожалуйста написать как это делается?

AleksandrIlinskii replied:
Твой код должен выглядеть вот так 
`backButton.setOnClickListener{ finish() }`

AleksandrIlinskii replied:
Почитай статью про стек активити https://metanit.com/java/android/2.10.php

sobolevkir replied:
Про finish() понял, но не понял про это: 

> Чтобы воспроизвести проблему, сделай вот как - с главной нажми на настройки. Потом на твою стрелку назад. Потом снова на настройки, снова назад и на настройки. А теперь нажимай не на стрелку назад, а на системную кнопку назад. Попадешь на главную. Но что должно произойти когда ты нажмешь назад? Правильно, должно выйти. А что будет? Ты попадешь на настройки, потому что они последующие в стеке

Почему с настроек должно было выйти, если на активити настроек мы можем попасть только с главной страницы?

AleksandrIlinskii replied:
Мы должны выйти с главной в систему, а не в настройки

AleksandrIlinskii commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,112 @@
+        android:layout_height="56dp"
+        android:gravity="center_vertical"
+        android:orientation="horizontal"
+        android:padding="4dp">
```
У тебя вертикальный LinearLayout. Элементов возде заголовка больше нет и вложенный LinearLayout избыточен

sobolevkir replied:
А я сделал это специально, чтобы можно было потом общим стилем Header сделать, а в нем уже textview, imagebutton'ы всякие и тд... Думаешь лучше убрать?

AleksandrIlinskii replied:
Ты же стиль применяешь на элемент (кнопка, текст, свитч и тд), а не на LinearLayout.

sobolevkir replied:
то есть LinearLayout нужен как заглушка со своим весом, чтобы отступ был?
Можешь поконкретнее сказать, что именно имеется в виду?

AleksandrIlinskii commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,112 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:layout_margin="16dp"
+        android:orientation="vertical">
```
Опять та ки - у тебя просто 3 кнопки подряд. Но тут скорее нужен вложенный LinearLayout, чем нет, чтобы можно было параметром layout_weight выставить вес кнопки и грамотно равномерно расположить внутри экрана

sobolevkir replied:
Так weight же задан в самих кнопках, для чего тогда делать ещё лэйаут?

AleksandrIlinskii replied:
Здесь он нужен, что бы отделить кнопки от header и расположить равномерно по высоте

sobolevkir replied:
Но тут же уже есть лэйаут на эти 3 кнопки? Нужен ещё какой-то здесь разве?
Просто смотри - я делаю общий лэйаут для 3х кнопок, делаю его высоту match_parent, а внутри него проставляю веса кнопок layout_weight="1". Таким образом общий лэйаут будет во всю оставшуюся высоту как раз. А вот если делать общий лэйаут так же через вес layout_weight="1", то студия выдаёт предупреждение, что вложенные веса могут уменьшить скорость работы.

AleksandrIlinskii replied:
Тут он как раз нужен. Я и пишу, что выше он избыточен, а тут необходим

AleksandrIlinskii commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,112 @@
+            android:id="@+id/button_search"
+            android:layout_width="match_parent"
+            android:layout_height="0dp"
+            android:layout_marginVertical="8dp"
```
Значения, которые повторяются несколько раз нужно выносить в ресурсы. Это критическое замечание

AleksandrIlinskii commented on app/src/main/res/values/colors.xml:
```diff
@@ -1,5 +1,9 @@
+    <color name="white">#ffffff</color>
+    <color name="dark">#1A1B22</color>
+    <color name="blue_background">#3772E7</color>
+    <color name="transparent">#00000000</color>
```
А чем отличается transparent от black?

sobolevkir replied:
а разве последние 2 цифры это не альфа?

AleksandrIlinskii replied:
С черным и белым это не работает
https://colorpicker.me

sobolevkir replied:
а как сделать прозрачный фон?

AleksandrIlinskii replied:
За этим к наставникам

AleksandrIlinskii commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,11 @@
+    <string name="share">Поделиться приложением</string>
+    <string name="support">Написать в поддержку</string>
+    <string name="user_agreement">Пользовательское соглашение</string>
+    <string name="back">Назад</string>
```
А где используется строка Назад?

sobolevkir replied:
Вот здесь:
android:contentDescription="@string/back"

Почему-то студия запрашивает наличие этого параметра для ImageView. Можешь пояснить, для чего он нужен?

AleksandrIlinskii replied:
Для людей с нарушениями зрения android система будет проговаривать что это за элемент. Можно просто поставить "" для заглушки

sobolevkir replied:
Может тогда лучше оставить эти описания?

AleksandrIlinskii replied:
В большом приложении будет слишком много лишних строк. Если нет требований на accessebility - то и не требуется делать. 
Есть принцип [YAGNI](https://ru.wikipedia.org/wiki/YAGNI), почитай о нем

sobolevkir committed:
Changes:
- when you click the "Back" button on the settings screen, now there is no creation of a new activity, but a return to the previous one in the stack;
- all duplicate sizes are taken out in resources, even margins and paddings;
- removed the extra layout on the main page.
